syft/formats/common/cyclonedxhelpers/decoder.go

```diff
@@ -92,9 +92,11 @@ func collectPackages(component *cyclonedx.Component, s *sbom.SBOM, idMap map[str
 		syftID := extractSyftPacakgeID(component.BOMRef)
 		if syftID != "" {
 			idMap[syftID] = p
+			p.OverrideID(artifact.ID(syftID))
+		} else {
+			// TODO there must be a better way than needing to call this manually:
+			p.SetID()
 		}
-		// TODO there must be a better way than needing to call this manually:
-		p.SetID()
```

Comment:
The only way this `p.OverrideID` gets called is if the `component.BOMRef` is a PURL that contains the `package-id` qualifier (which is probably only set when Syft generated the SBOM in the first place, using a PURL as the BOMRef). Additionally, this `syftID` might not even match the ID generated by the _current version of Syft_, so the way it's being used is probably wrong.

It was originally a conscious decision to re-generate the IDs for components (by calling the `p.SetID()` instead of reusing whatever ID is provided by the format in order to get reasonably consistent output. For example: if we used the BOMRef as the Syft package ID, then re-encoded CycloneDX, we'd have something like: `pkg:thing/name@version?package-id=pkg:thing/name@version?package-id=2a7443b435345b2` (ignore the lack of URL encoding). Obviously we can handle this during encoding, for example, by comparing the ID to what `SetID` would set.

A few things here:
* we should probably get rid of `extractSyftPacakgeID`. I don't think this is providing any value -- at best it's extracting the same ID that `SetID()` would generate, but more likely it is setting some different ID that is not useful in any way
* we should adjust output formats to be aware when an ID is explicitly set vs generated based on the data (this would probably look like 1 or 2 utility methods on the package -- maybe a `GenerateID` that could be used for comparison and/or an `IsIDGenerated()`)
  * We would also need to update the CycloneDX encoder and possibly other encoders to handle cases with explicitly set IDs

If we want to support reading IDs from formats instead of always generating them based on the data (I'm not sure we always do -- this seems like a configurable option), I think your observation that we should be using the `BOMRef` is accurate, maybe something like this:
```go
		if component.BOMRef != "" {
			p.OverrideID(artifact.ID(component.BOMRef))
		} else {
			p.SetID()
		}
		idMap[string(p.ID())] = p
```